Adds section on product quantization for docs (#6926)
opensearch-project/documentation-website: _search-plugins/knn/knn-vector-quantization.md

# k-NN vector quantization

> By default, the k-NN plugin supports the indexing and querying of vectors of type `float`, where each dimension of the 

Author: I made the line breaks so that editing would be easier and it doesnt impact rendering (i.e. it wouldnt be one line that rolls out of the screen). Is this incorrect to do?

> By default, the k-NN plugin supports the indexing and querying of vectors of type `float`, where each dimension of the 

Reviewer: Yes, it's incorrect to enter line breaks. The site and OpenSearch Project doc team follow a specific formatting guide.

By default, the k-NN plugin supports the indexing and querying of vectors of type `float`, where each dimension of the vector occupies 4 bytes of memory. For use cases that require ingestion on a large scale, keeping `float` vectors can be expensive because OpenSearch needs to construct, load, save, and search graphs (for native `nmslib` and `faiss` engines). To reduce the memory footprint, you can use vector quantization.

> In OpenSearch, many varieties of quantization are supported. In general, the level of quantization will provide a trade-off between the accuracy of the nearest neighbor search and the size of the memory footprint the vector search system will consume. The supported types include byte vectors, 16-bit scalar quantization, and product quantization (PQ).

Reviewer 2: ```suggestion
OpenSearch supports many varieties of quantization. In general, the level of quantization will provide a trade-off between the accuracy of the nearest neighbor search and the size of the memory footprint consumed by the vector search. The supported types include byte vectors, 16-bit scalar quantization, and product quantization (PQ).
```

Optionally, you can specify the parameters in `method.parameters.encoder`. For more information about `encoder` object parameters, see [SQ parameters]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index/#sq-parameters).

The `fp16` encoder converts 32-bit vectors into their 16-bit counterparts. For this encoder type, the vector values must be in the [-65504.0, 65504.0] range. To define how to handle out-of-range values, the preceding request specifies the `clip` parameter. By default, this parameter is `false`, and any vectors containing out-of-range values are rejected. 

When `clip` is set to `true` (as in the preceding request), out-of-range vector values are rounded up or down so that they are in the supported range. For example, if the original 32-bit vector is `[65510.82, -65504.1]`, the vector will be indexed as a 16-bit vector `[65504.0, -65504.0]`.

We recommend setting `clip` to `true` only if very few elements lie outside of the supported range. Rounding the values may cause a drop in recall.
{: .note}

> The following example method definition specifies the Faiss SQfp16 encoder, which rejects any indexing request that contains out-of-range vector values (because the `clip` parameter is `false` by default).

Reviewer 2: Shouldn't this end in a colon because it introduces an example?

> The following example method definition specifies the Faiss SQfp16 encoder, which rejects any indexing request that contains out-of-range vector values (because the `clip` parameter is `false` by default).

Reviewer: ```suggestion
The following example method definition specifies the Faiss SQfp16 encoder, which rejects any indexing request that contains out-of-range vector values (because the `clip` parameter is `false` by default):
```

```
{% include copy-curl.html %}

> During ingestion, make sure each dimension of the vector is in the supported range ([-65504.0, 65504.0]).

Reviewer 2: ```suggestion
During ingestion, make sure each vector dimension is in the supported range ([-65504.0, 65504.0]).
```

> During ingestion, make sure each dimension of the vector is in the supported range ([-65504.0, 65504.0]).

Reviewer 2: Shouldn't this end in a colon because it introduces an example?

```
{% include copy-curl.html %}

> During querying, the query vector has no range limitation.

Reviewer 2: Shouldn't this end in a colon because it introduces an example?

> During querying, the query vector has no range limitation.

Reviewer: ```suggestion
During querying, the query vector has no range limitation:
```

> During querying, there is no range limitation for the query vector.

Reviewer: ```suggestion
During querying, the query vector has no range limitation.
```

1.1 * (((2 * 256) * 1,000,000) + (4 * 128 * 256))  ~= 0.525 GB
```

## Faiss product quantization

> PQ is a technique used to represent a vector in a configurable amount of bits. In general, it can be used to achieve a higher level of compression compared to byte and scalar quantization. PQ works by breaking up vectors into _m_ subvectors and encoding each subvector with _code_size_ bits. Thus, the total amount of memory for the vector ends up being `m*code_size` bits, plus overhead. For details about the parameters, see [PQ parameters]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index/#pq-parameters). PQ is only supported for the _Faiss_ engine and can be used with either the _HNSW_ or the _IVF_ ANN (Approximate Nearest Neighbor) algorithms.

Reviewer 2: ```suggestion
PQ is a technique used to represent a vector in a configurable amount of bits. In general, it can be used to achieve a higher level of compression as compared to byte or scalar quantization. PQ works by separating vectors into _m_ subvectors and encoding each subvector with _code_size_ bits. Thus, the total amount of memory for the vector is `m*code_size` bits, plus overhead. For details about the parameters, see [PQ parameters]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index/#pq-parameters). PQ is only supported for the _Faiss_ engine and can be used with either the _HNSW_ or _IVF_ approximate nearest neighbor (ANN) algorithms.
```

> Product quantization is a technique that allows users to represent a vector in a configurable amount of bits. In general, it can be used to achieve a higher level of compression compared to byte and scalar quantization. Product quantization works by breaking up vectors into _m_ subvectors, and  encoding each subvector with _code_size_ bits. Thus, the total amount of memory for the vector ends up being `m*code_size` bits, plus overhead. For more details about the parameters of product quantization, see [PQ parameters]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index/#pq-parameters). Product quantization is only supported for the _Faiss_ engine and can be used with either the _HNSW_ or the _IVF_ ANN algorithms.

Reviewer: ```suggestion
PQ is a technique used to represent a vector in a configurable amount of bits. In general, it can be used to achieve a higher level of compression compared to byte and scalar quantization. PQ works by breaking up vectors into _m_ subvectors and  encoding each subvector with _code_size_ bits. Thus, the total amount of memory for the vector ends up being `m*code_size` bits, plus overhead. For details about the parameters, go to [PQ parameters]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index/#pq-parameters). PQ is only supported for the _Faiss_ engine and can be used with either the _HNSW_ or the _IVF_ ANN (Artificial Neural Network) algorithms.
```

Author: ANN stands for Approxximate Nearest Neighbor. Ill fix

### Using Faiss product quantization

> To minimize the loss in accuracy, PQ requires a _training_ step that builds a model based on the distribution of the data that will be searched over.

Reviewer 2: ```suggestion
To minimize loss in accuracy, PQ requires a _training_ step that builds a model based on the distribution of the data that will be searched.
```

> In order to minimize the loss in accuracy, product quantization requires a _training_ step that builds a model based on the distribution of the data that will be searched over.

Reviewer: ```suggestion
To minimize the loss in accuracy, PQ requires a _training_ step that builds a model based on the distribution of the data that will be searched over.
```

> The product quantizer is trained by running k-means clustering on a set of training vectors for each sub-vector space and extracts the centroids to be used for the encoding. The training vectors can be either a subset of the vectors to be ingested or vectors that have the same distribution and dimension as the vectors to be ingested.

Reviewer 2: ```suggestion
The product quantizer is trained by running k-means clustering on a set of training vectors for each subvector space and extracts the centroids to be used for encoding. The training vectors can be either a subset of the vectors to be ingested or vectors that have the same distribution and dimension as the vectors to be ingested.
```

> Under the hood, the product quantizer is trained by running k-Means clustering on a set of training vectors for each sub-vector space and extracts the centroids to be used for the encoding. The training vectors can either be a subset of the vectors to be ingested, or vectors that have the same distribution and dimension as the vectors to be ingested.

Reviewer: Avoid using jargon.

Author: What do you mean by jargon?

Reviewer: For example, "under the hood" is jargon. Please see the OpenSearch writing style guide about tone (https://github.com/opensearch-project/documentation-website/blob/main/STYLE_GUIDE.md#tone). To move this forward, I've accepted writing and editorial changes based on the style guide. I'll move this forward into editorial and once approved get the PR merged. Thank you for contributing to the documentation :)

> Under the hood, the product quantizer is trained by running k-Means clustering on a set of training vectors for each sub-vector space and extracts the centroids to be used for the encoding. The training vectors can either be a subset of the vectors to be ingested, or vectors that have the same distribution and dimension as the vectors to be ingested.

Reviewer: ```suggestion
The product quantizer is trained by running k-means clustering on a set of training vectors for each sub-vector space and extracts the centroids to be used for the encoding. The training vectors can be either a subset of the vectors to be ingested or vectors that have the same distribution and dimension as the vectors to be ingested.
```

> In OpenSearch, the training vectors need to be present in an index. In general, the amount of training data will depend on which ANN algorithm will be used and how much data will go into the index. For IVF-based indices, a good number of training vectors to use is `max(1000*nlist, 2^code_size * 1000)`. For HNSW-based indexes, a good number is `2^code_size*1000` training vectors. See [Faiss's documentation](https://github.com/facebookresearch/faiss/wiki/FAQ#how-many-training-points-do-i-need-for-k-means) for more details about the methodology behind calculating these figures.

Reviewer 2: ```suggestion
In OpenSearch, the training vectors need to be present in an index. In general, the amount of training data will depend on which ANN algorithm is used and how much data will be stored in the index. For IVF-based indexes, a recommended number of training vectors is `max(1000*nlist, 2^code_size * 1000)`. For HNSW-based indexes, a recommended number is `2^code_size*1000`. See the [Faiss documentation](https://github.com/facebookresearch/faiss/wiki/FAQ#how-many-training-points-do-i-need-for-k-means) for more information about the methodology used to calculate these figures.
```

> In OpenSearch, the training vectors need to be present in an index. In general, the amount of training data will depend on which ANN algorithm will be used and how much data will go into the index. For IVF-based indices, a good number of training vectors to use is `max(1000*nlist, 2^code_size * 1000)`. For HNSW-based indices, a good number is `2^code_size*1000` training vectors. See [Faiss's documentation](https://github.com/facebookresearch/faiss/wiki/FAQ#how-many-training-points-do-i-need-for-k-means) for more details on how these numbers are arrived at.

Reviewer: ```suggestion
In OpenSearch, the training vectors need to be present in an index. In general, the amount of training data will depend on which ANN algorithm will be used and how much data will go into the index. For IVF-based indices, a good number of training vectors to use is `max(1000*nlist, 2^code_size * 1000)`. For HNSW-based indices, a good number is `2^code_size*1000` training vectors. Go to [Faiss's documentation](https://github.com/facebookresearch/faiss/wiki/FAQ#how-many-training-points-do-i-need-for-k-means) for more details about the methodology behind calculating these figures.
```

> For PQ, the two parameters that need to be selected are _m_ and _code_size_. _m_ determines how many sub-vectors the vectors should be split to encode separately. Consequently, the _dimension_ needs to be divisible by _m_. _code_size_ determines how many bits each sub-vector will be encoded with. In general, a good place to start is setting `code_size = 8` and then tuning _m_ to get the desired trade-off between memory footprint and recall.

Reviewer 2: ```suggestion
For PQ, both _m_ and _code_size_ need to be selected. _m_ determines how many subvectors the vectors should be split to encode separately. Consequently, the _dimension_ needs to be divisible by _m_. _code_size_ determines how many bits each sub-vector will be encoded with. In general, we recommend a setting of `code_size = 8` and then tuning _m_ to get the desired trade-off between memory footprint and recall.
```

> For PQ, the two parameters that need to be selected are _m_ and _code_size_. _m_ determines how many sub-vectors the vectors should be split to encode separately. Consequently, the _dimension_ needs to be divisible by _m_. _code_size_ determines how many bits each sub-vector will be encoded with. In general, a good place to start is setting `code_size = 8` and then tuning _m_ to get the desired trade-off between memory footprint and recall.

Reviewer 2: I'm not following the second sentence here. Do we mean something like "_m_ determines the number of subvectors into which vectors should be split for separate encoding"? In the fourth sentence, is "with" the correct preposition, or should it be "into"?

Reviewer: Yes, your rewrite is correct. I revised the following sentence to read: `_code_size_ determines the number of bits used to encode each subvector.`

> For product quantization, the two parameters that need to be selected are _m_ and _code_size_. _m_ determines how many sub-vectors the vectors should be broken up into to encode separately - consequently, the _dimension_ needs to be divisible by _m_. _code_size_ determines how many bits each sub-vector will be encoded with. In general, a good place to start is setting `code_size = 8` and then tuning _m_ to get the desired tradeoff between memory footprint and recall.

Reviewer: ```suggestion
For PQ, the two parameters that need to be selected are _m_ and _code_size_. _m_ determines how many sub-vectors the vectors should be split into to encode separately. Consequently, the _dimension_ needs to be divisible by _m_. _code_size_ determines how many bits each sub-vector will be encoded with. In general, a good place to start is setting `code_size = 8` and then tuning _m_ to get the desired trade-off between memory footprint and recall.
```

For an example of setting up an index with PQ, see the [Building a k-NN index from a model]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/#building-a-k-nn-index-from-a-model) tutorial.

> ### Memory Estimation

Reviewer 2: ```suggestion
### Memory estimation
```

> For an example of setting up an index with product quantization, see [this tutorial]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/#building-a-k-nn-index-from-a-model).

Reviewer: ```suggestion
For an example of setting up an index with PQ, go to the [Building a k-NN index from a model]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/#building-a-k-nn-index-from-a-model) tutorial.
```

> While PQ is meant to represent individual vectors with `m*code_size` bits, in reality the indexes take up more space. This is mainly due to the overhead of storing certain code tables and auxiliary data structures.

Reviewer 2: ```suggestion
While PQ is meant to represent individual vectors with `m*code_size` bits, in reality, the indexes consume more space. This is mainly due to the overhead of storing certain code tables and auxiliary data structures.
```

> While product quantization is meant to represent individual vectors with `m*code_size` bits, in reality the indices take up more space than this. This is mainly due to the overhead of storing certain code tables and auxilary data structures.

Reviewer: ```suggestion
While PQ is meant to represent individual vectors with `m*code_size` bits, in reality the indexes take up more space. This is mainly due to the overhead of storing certain code tables and auxiliary data structures.
```

> Some of the memory formulas depend on the number of segments present. Typically, this is not known beforehand but a good default value is 300.

Reviewer 2: ```suggestion
Some of the memory formulas depend on the number of segments present. This is not typically known beforehand, but a recommended default value is 300.
```

{: .note}

#### HNSW memory estimation

The memory required for HNSW with PQ is estimated to be `1.1*(((pq_code_size / 8) * pq_m + 24 + 8 * hnsw_m) * num_vectors + num_segments * (2^pq_code_size * 4 * d))` bytes.

> As an example, assume that you have 1 million vectors with a dimension of 256, `hnsw_m` of 16, `pq_m` of 32, `pq_code_size` of 8 and 100 segments. The memory requirement can be estimated as follows:

Reviewer: ```suggestion
As an example, assume that you have 1 million vectors with a dimension of 256, `hnsw_m` of 16, `pq_m` of 32, `pq_code_size` of 8, and 100 segments. The memory requirement can be estimated as follows:
```

> As an example, assume that you have 1 million vectors with a dimension of 256, `hnsw_m` of 16, `pq_m` of 32, `pq_code_size` of 8 and 100 segments. The memory requirement can be estimated as follows:

Reviewer: ```suggestion
For example, assume that you have 1 million vectors with a dimension of 256, `hnsw_m` of 16, `pq_m` of 32, `pq_code_size` of 8 and 100 segments. The memory requirement can be estimated as follows:
```

```bash
1.1*((8 / 8 * 32 + 24 + 8 * 16) * 1000000 + 100 * (2^8 * 4 * 256)) ~= 0.215 GB
```

#### IVF memory estimation

The memory required for IVF with PQ is estimated to be `1.1*(((pq_code_size / 8) * pq_m + 24) * num_vectors  + num_segments * (2^code_size * 4 * d + 4 * ivf_nlist * d))` bytes.

> For example, assume that you have 1 million vectors with a dimension of 256, `ivf_nlist` of 512, `pq_m` of 32, `pq_code_size` of 8 and 100 segments. The memory requirement can be estimated as follows:

Reviewer 2: ```suggestion
For example, assume that you have 1 million vectors with a dimension of 256, `ivf_nlist` of 512, `pq_m` of 32, `pq_code_size` of 8, and 100 segments. The memory requirement can be estimated as follows:
```

> As an example, assume that you have 1 million vectors with a dimension of 256, `ivf_nlist` of 512, `pq_m` of 32, `pq_code_size` of 8 and 100 segments. The memory requirement can be estimated as follows:

Reviewer: ```suggestion
For example, assume that you have 1 million vectors with a dimension of 256, `ivf_nlist` of 512, `pq_m` of 32, `pq_code_size` of 8 and 100 segments. The memory requirement can be estimated as follows:
```

for more details on how these numbers are arrived at.

For product quantization, the two parameters that need to be selected are _m_ and _code_size_. _m_ determines how many 
sub-vectors the vectors should be broken up into to encode separately - consequently, the _dimension_ needs to be 
divisible by _m_. _code_size_ determines how many bits each sub-vector will be encoded with. In general, a good place to
start is setting `code_size = 8` and then tuning _m_ to get the desired tradeoff between memory footprint and recall.

For an example of setting up an index with product quantization, see [this tutorial]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/#building-a-k-nn-index-from-a-model).

### Memory Estimation

While product quantization is meant to represent individual vectors with `m*code_size` bits, in reality the indices 
take up more space than this. This is mainly due to the overhead of storing certain code tables and auxilary data 
structures.

Some of the memory formulas depend on the number of segments present. Typically, this is not known beforehand but a good
default value is 300.
{: .note}

#### HNSW memory estimation

> The memory required for HNSW with PQ is estimated to be `1.1*((pq_code_size / 8 * pq_m + 24 + 8 * hnsw_m) * num_vectors + num_segments * (2^pq_code_size * 4 * d))` bytes.

Reviewer 3: Should we add parenthesis for each component to avoid operator precedence issues for evaluating memory? For example (pq_code_size/8)*pq_m

Author: I think it makes sense with respect to the "/" operation. For others, it may lead to too many parenthesis. I updated
